### ISSUE TITLE
Refactor tax calculation to pull from AvaTax line-level detail instead of summary-level detail

### ISSUE THREAD

**Issue by LordZardeck** [classyllama/ClassyLlama_AvaTax#196]
Currently, the Magento extension uses the "summary" data from the AvaTax response to calculate taxes in Magento. However with the upcoming release of Customs Duty & Import Tax, we're needing to refactor the extension to pull tax details from the "lines > details" section of the JSON response.

Currently, the "REPORTS > Sales > Tax" reports in the Magento admin is completely useless for calculating AvaTax taxes, however after this refactor the report should be more accurate. However we recommend that users still pull tax reports from Avalara as the accuracy of the Magento report has not been fully verified.

Once this change is made, we will be reverting the changes made in the #90 PR that resolved issue #70 as that code will no longer be necessary.

As a point of reference, here is a JSON response from AvaTax:

```
{
    "id": 529008084,
    "code": "99608ccb-4415-416a-ab60-8e498f560a40",
    "companyId": '<REDACTED>',
    "date": "2018-03-09",
    "status": "Saved",
    "type": "SalesInvoice",
    "batchCode": "",
    "currencyCode": "CAD",
    "customerUsageType": "",
    "entityUseCode": "",
    "customerVendorCode": "ABC",
    "customerCode": "ABC",
    "exemptNo": "",
    "reconciled": false,
    "purchaseOrderNo": "109288",
    "referenceCode": "",
    "salespersonCode": "",
    "taxOverrideType": "None",
    "taxOverrideAmount": 0,
    "taxOverrideReason": "",
    "totalAmount": 1100,
    "totalExempt": 0,
    "totalDiscount": 0,
    "totalTax": 2122.25,
    "totalTaxable": 2985,
    "totalTaxCalculated": 2122.25,
    "adjustmentReason": "NotAdjusted",
    "adjustmentDescription": "",
    "locked": false,
    "region": "ON",
    "country": "CA",
    "version": 1,
    "softwareVersion": "18.10.0.21",
    "originAddressId": 1353014510,
    "destinationAddressId": 1353014511,
    "exchangeRateEffectiveDate": "2018-03-09",
    "exchangeRate": 1,
    "isSellerImporterOfRecord": true,
    "businessIdentificationNo": "",
    "modifiedDate": "2018-10-19T16:36:42.17",
    "modifiedUserId": 309923,
    "taxDate": "2018-03-09T00:00:00",
    "lines": [
        {
            "id": 1944483755,
            "transactionId": 529008084,
            "lineNumber": "1",
            "customerUsageType": "",
            "entityUseCode": "",
            "description": "",
            "destinationAddressId": 1353014511,
            "originAddressId": 1353014510,
            "discountAmount": 0,
            "exemptAmount": 0,
            "exemptCertId": 0,
            "exemptNo": "",
            "isItemTaxable": true,
            "isSSTP": false,
            "itemCode": "",
            "lineAmount": 1000,
            "quantity": 1,
            "ref1": "",
            "ref2": "",
            "reportingDate": "2018-03-09",
            "revAccount": "",
            "sourcing": "Destination",
            "tax": 2109.25,
            "taxableAmount": 2885,
            "taxCalculated": 2109.25,
            "taxCode": "P0000000",
            "taxCodeId": 8087,
            "taxDate": "2018-03-09",
            "taxEngine": "",
            "taxOverrideType": "None",
            "businessIdentificationNo": "",
            "taxOverrideAmount": 0,
            "taxOverrideReason": "",
            "taxIncluded": false,
            "details": [
                {
                    "id": 5309476423,
                    "transactionLineId": 1944483755,
                    "transactionId": 529008084,
                    "addressId": 1353014511,
                    "country": "CA",
                    "region": "CA",
                    "stateFIPS": "CA",
                    "exemptAmount": 0,
                    "exemptReasonId": 4,
                    "inState": false,
                    "jurisCode": "CA",
                    "jurisName": "CANADA",
                    "jurisdictionId": 20451988,
                    "stateAssignedNo": "",
                    "jurisType": "CNT",
                    "jurisdictionType": "Country",
                    "nonTaxableAmount": 0,
                    "nonTaxableRuleId": 0,
                    "nonTaxableType": "RateRule",
                    "rate": 0.05,
                    "rateRuleId": 322399,
                    "rateSourceId": 0,
                    "serCode": "",
                    "sourcing": "Destination",
                    "tax": 144.25,
                    "taxableAmount": 2885,
                    "taxType": "Output",
                    "taxSubTypeId": "O",
                    "taxTypeGroupId": "InputAndOutput",
                    "taxName": "CANADA GST/TPS",
                    "taxAuthorityTypeId": 45,
                    "taxRegionId": 0,
                    "taxCalculated": 144.25,
                    "taxOverride": 0,
                    "rateType": "Standard",
                    "rateTypeCode": "S",
                    "taxableUnits": 2885,
                    "nonTaxableUnits": 0,
                    "exemptUnits": 0,
                    "unitOfBasis": "PerCurrencyUnit",
                    "isNonPassThru": false
                },
                {
                    "id": 5309476425,
                    "transactionLineId": 1944483755,
                    "transactionId": 529008084,
                    "addressId": 1353014511,
                    "country": "CA",
                    "region": "CA",
                    "stateFIPS": "CA",
                    "exemptAmount": 0,
                    "exemptReasonId": 4,
                    "inState": false,
                    "jurisCode": "CA",
                    "jurisName": "CANADA",
                    "jurisdictionId": 20451988,
                    "stateAssignedNo": "",
                    "jurisType": "CNT",
                    "jurisdictionType": "Country",
                    "nonTaxableAmount": 0,
                    "nonTaxableRuleId": 0,
                    "nonTaxableType": "RateRule",
                    "rate": 188.5,
                    "rateRuleId": 3546260,
                    "rateSourceId": 4,
                    "serCode": "",
                    "sourcing": "Destination",
                    "tax": 1885,
                    "taxableAmount": 0,
                    "taxType": "LandedCost",
                    "taxSubTypeId": "ImportDuty",
                    "taxTypeGroupId": "LandedCost",
                    "taxName": "CA LandedCost",
                    "taxAuthorityTypeId": 45,
                    "taxRegionId": 0,
                    "taxCalculated": 1885,
                    "taxOverride": 0,
                    "rateType": "Standard",
                    "rateTypeCode": "S",
                    "taxableUnits": 10,
                    "nonTaxableUnits": 0,
                    "exemptUnits": 0,
                    "unitOfBasis": "Mass",
                    "isNonPassThru": false
                },
                {
                    "id": 5309476424,
                    "transactionLineId": 1944483755,
                    "transactionId": 529008084,
                    "addressId": 1353014511,
                    "country": "CA",
                    "region": "ON",
                    "stateFIPS": "ON",
                    "exemptAmount": 0,
                    "exemptReasonId": 4,
                    "inState": false,
                    "jurisCode": "ON",
                    "jurisName": "ONTARIO",
                    "jurisdictionId": 20451997,
                    "stateAssignedNo": "",
                    "jurisType": "STA",
                    "jurisdictionType": "State",
                    "nonTaxableAmount": 0,
                    "nonTaxableRuleId": 0,
                    "nonTaxableType": "RateRule",
                    "rate": 0.08,
                    "rateRuleId": 394650,
                    "rateSourceId": 0,
                    "serCode": "",
                    "sourcing": "Destination",
                    "tax": 80,
                    "taxableAmount": 1000,
                    "taxType": "Output",
                    "taxSubTypeId": "O",
                    "taxTypeGroupId": "InputAndOutput",
                    "taxName": "ONTARIO HST",
                    "taxAuthorityTypeId": 45,
                    "taxRegionId": 0,
                    "taxCalculated": 80,
                    "taxOverride": 0,
                    "rateType": "Standard",
                    "rateTypeCode": "S",
                    "taxableUnits": 1000,
                    "nonTaxableUnits": 0,
                    "exemptUnits": 0,
                    "unitOfBasis": "PerCurrencyUnit",
                    "isNonPassThru": false
                }
            ],
            "nonPassthroughDetails": [],
            "lineLocationTypes": [
                {
                    "documentLineLocationTypeId": 4151373982,
                    "documentLineId": 1944483755,
                    "documentAddressId": 1353014511,
                    "locationTypeCode": "ShipTo"
                },
                {
                    "documentLineLocationTypeId": 4151373981,
                    "documentLineId": 1944483755,
                    "documentAddressId": 1353014510,
                    "locationTypeCode": "ShipFrom"
                }
            ],
            "parameters": {
                "Mass.UOM": "Kilogram",
                "AvaTax.LandedCost.UnitName": "Kilogram",
                "AvaTax.LandedCost.UnitAmount": "10000",
                "Mass": "10000",
                "DeminimisMet": "True",
                "AvaTax.LandedCost.HTSCode": "1108191200"
            },
            "hsCode": "1108191200",
            "vatCode": "",
            "vatNumberTypeId": 0
        },
        {
            "id": 1944483756,
            "transactionId": 529008084,
            "lineNumber": "2",
            "boundaryOverrideId": 0,
            "customerUsageType": "",
            "entityUseCode": "",
            "description": "Shipping",
            "destinationAddressId": 1353014511,
            "originAddressId": 1353014510,
            "discountAmount": 0,
            "exemptAmount": 0,
            "exemptCertId": 0,
            "exemptNo": "",
            "isItemTaxable": true,
            "isSSTP": false,
            "itemCode": "",
            "lineAmount": 50,
            "quantity": 1,
            "ref1": "",
            "ref2": "",
            "reportingDate": "2018-03-09",
            "revAccount": "",
            "sourcing": "Destination",
            "tax": 6.5,
            "taxableAmount": 50,
            "taxCalculated": 6.5,
            "taxCode": "FR010000",
            "taxCodeId": 8551,
            "taxDate": "2018-03-09",
            "taxEngine": "",
            "taxOverrideType": "None",
            "businessIdentificationNo": "",
            "taxOverrideAmount": 0,
            "taxOverrideReason": "",
            "taxIncluded": false,
            "details": [
                {
                    "id": 5309476426,
                    "transactionLineId": 1944483756,
                    "transactionId": 529008084,
                    "addressId": 1353014511,
                    "country": "CA",
                    "region": "CA",
                    "stateFIPS": "CA",
                    "exemptAmount": 0,
                    "exemptReasonId": 4,
                    "inState": false,
                    "jurisCode": "CA",
                    "jurisName": "CANADA",
                    "jurisdictionId": 20451988,
                    "stateAssignedNo": "",
                    "jurisType": "CNT",
                    "jurisdictionType": "Country",
                    "nonTaxableAmount": 0,
                    "nonTaxableRuleId": 0,
                    "nonTaxableType": "RateRule",
                    "rate": 0.05,
                    "rateRuleId": 322399,
                    "rateSourceId": 0,
                    "serCode": "",
                    "sourcing": "Destination",
                    "tax": 2.5,
                    "taxableAmount": 50,
                    "taxType": "Output",
                    "taxSubTypeId": "O",
                    "taxTypeGroupId": "InputAndOutput",
                    "taxName": "CANADA GST/TPS",
                    "taxAuthorityTypeId": 45,
                    "taxRegionId": 0,
                    "taxCalculated": 2.5,
                    "taxOverride": 0,
                    "rateType": "Standard",
                    "rateTypeCode": "S",
                    "taxableUnits": 50,
                    "nonTaxableUnits": 0,
                    "exemptUnits": 0,
                    "unitOfBasis": "PerCurrencyUnit",
                    "isNonPassThru": false
                },
                {
                    "id": 5309476427,
                    "transactionLineId": 1944483756,
                    "transactionId": 529008084,
                    "addressId": 1353014511,
                    "country": "CA",
                    "region": "ON",
                    "stateFIPS": "ON",
                    "exemptAmount": 0,
                    "exemptReasonId": 4,
                    "inState": false,
                    "jurisCode": "ON",
                    "jurisName": "ONTARIO",
                    "jurisdictionId": 20451997,
                    "stateAssignedNo": "",
                    "jurisType": "STA",
                    "jurisdictionType": "State",
                    "nonTaxableAmount": 0,
                    "nonTaxableRuleId": 0,
                    "nonTaxableType": "RateRule",
                    "rate": 0.08,
                    "rateRuleId": 394650,
                    "rateSourceId": 0,
                    "serCode": "",
                    "sourcing": "Destination",
                    "tax": 4,
                    "taxableAmount": 50,
                    "taxType": "Output",
                    "taxSubTypeId": "O",
                    "taxTypeGroupId": "InputAndOutput",
                    "taxName": "ONTARIO HST",
                    "taxAuthorityTypeId": 45,
                    "taxRegionId": 0,
                    "taxCalculated": 4,
                    "taxOverride": 0,
                    "rateType": "Standard",
                    "rateTypeCode": "S",
                    "taxableUnits": 50,
                    "nonTaxableUnits": 0,
                    "exemptUnits": 0,
                    "unitOfBasis": "PerCurrencyUnit",
                    "isNonPassThru": false
                }
            ],
            "nonPassthroughDetails": [],
            "lineLocationTypes": [
                {
                    "documentLineLocationTypeId": 4151373984,
                    "documentLineId": 1944483756,
                    "documentAddressId": 1353014511,
                    "locationTypeCode": "ShipTo"
                },
                {
                    "documentLineLocationTypeId": 4151373983,
                    "documentLineId": 1944483756,
                    "documentAddressId": 1353014510,
                    "locationTypeCode": "ShipFrom"
                }
            ],
            "parameters": {},
            "hsCode": "",
            "vatCode": "",
            "vatNumberTypeId": 0
        },
        {
            "id": 1944483757,
            "transactionId": 529008084,
            "lineNumber": "3",
            "boundaryOverrideId": 0,
            "customerUsageType": "",
            "entityUseCode": "",
            "description": "Insurance",
            "destinationAddressId": 1353014511,
            "originAddressId": 1353014510,
            "discountAmount": 0,
            "exemptAmount": 0,
            "exemptCertId": 0,
            "exemptNo": "",
            "isItemTaxable": true,
            "isSSTP": false,
            "itemCode": "",
            "lineAmount": 50,
            "quantity": 1,
            "ref1": "",
            "ref2": "",
            "reportingDate": "2018-03-09",
            "revAccount": "",
            "sourcing": "Destination",
            "tax": 6.5,
            "taxableAmount": 50,
            "taxCalculated": 6.5,
            "taxCode": "FR070100",
            "taxCodeId": 52432,
            "taxDate": "2018-03-09",
            "taxEngine": "",
            "taxOverrideType": "None",
            "businessIdentificationNo": "",
            "taxOverrideAmount": 0,
            "taxOverrideReason": "",
            "taxIncluded": false,
            "details": [
                {
                    "id": 5309476428,
                    "transactionLineId": 1944483757,
                    "transactionId": 529008084,
                    "addressId": 1353014511,
                    "country": "CA",
                    "region": "CA",
                    "stateFIPS": "CA",
                    "exemptAmount": 0,
                    "exemptReasonId": 4,
                    "inState": false,
                    "jurisCode": "CA",
                    "jurisName": "CANADA",
                    "jurisdictionId": 20451988,
                    "stateAssignedNo": "",
                    "jurisType": "CNT",
                    "jurisdictionType": "Country",
                    "nonTaxableAmount": 0,
                    "nonTaxableRuleId": 0,
                    "nonTaxableType": "RateRule",
                    "rate": 0.05,
                    "rateRuleId": 322399,
                    "rateSourceId": 0,
                    "serCode": "",
                    "sourcing": "Destination",
                    "tax": 2.5,
                    "taxableAmount": 50,
                    "taxType": "Output",
                    "taxSubTypeId": "O",
                    "taxTypeGroupId": "InputAndOutput",
                    "taxName": "CANADA GST/TPS",
                    "taxAuthorityTypeId": 45,
                    "taxRegionId": 0,
                    "taxCalculated": 2.5,
                    "taxOverride": 0,
                    "rateType": "Standard",
                    "rateTypeCode": "S",
                    "taxableUnits": 50,
                    "nonTaxableUnits": 0,
                    "exemptUnits": 0,
                    "unitOfBasis": "PerCurrencyUnit",
                    "isNonPassThru": false
                },
                {
                    "id": 5309476429,
                    "transactionLineId": 1944483757,
                    "transactionId": 529008084,
                    "addressId": 1353014511,
                    "country": "CA",
                    "region": "ON",
                    "stateFIPS": "ON",
                    "exemptAmount": 0,
                    "exemptReasonId": 4,
                    "inState": false,
                    "jurisCode": "ON",
                    "jurisName": "ONTARIO",
                    "jurisdictionId": 20451997,
                    "stateAssignedNo": "",
                    "jurisType": "STA",
                    "jurisdictionType": "State",
                    "nonTaxableAmount": 0,
                    "nonTaxableRuleId": 0,
                    "nonTaxableType": "RateRule",
                    "rate": 0.08,
                    "rateRuleId": 394650,
                    "rateSourceId": 0,
                    "serCode": "",
                    "sourcing": "Destination",
                    "tax": 4,
                    "taxableAmount": 50,
                    "taxType": "Output",
                    "taxSubTypeId": "O",
                    "taxTypeGroupId": "InputAndOutput",
                    "taxName": "ONTARIO HST",
                    "taxAuthorityTypeId": 45,
                    "taxRegionId": 0,
                    "taxCalculated": 4,
                    "taxOverride": 0,
                    "rateType": "Standard",
                    "rateTypeCode": "S",
                    "taxableUnits": 50,
                    "nonTaxableUnits": 0,
                    "exemptUnits": 0,
                    "unitOfBasis": "PerCurrencyUnit",
                    "isNonPassThru": false
                }
            ],
            "nonPassthroughDetails": [],
            "lineLocationTypes": [
                {
                    "documentLineLocationTypeId": 4151373986,
                    "documentLineId": 1944483757,
                    "documentAddressId": 1353014511,
                    "locationTypeCode": "ShipTo"
                },
                {
                    "documentLineLocationTypeId": 4151373985,
                    "documentLineId": 1944483757,
                    "documentAddressId": 1353014510,
                    "locationTypeCode": "ShipFrom"
                }
            ],
            "parameters": {},
            "hsCode": "",
            "vatCode": "",
            "vatNumberTypeId": 0
        }
    ],
    "addresses": [
        {
            "id": 1353014510,
            "transactionId": 529008084,
            "boundaryLevel": "Zip5",
            "line1": "",
            "line2": "",
            "line3": "",
            "city": "",
            "region": "",
            "postalCode": "",
            "country": "MX",
            "taxRegionId": 205098
        },
        {
            "id": 1353014511,
            "transactionId": 529008084,
            "boundaryLevel": "Zip5",
            "line1": "",
            "line2": "",
            "line3": "",
            "city": "9 McKinnon Ave",
            "region": "ON",
            "postalCode": "L7G 5H6",
            "country": "CA",
            "taxRegionId": 0
        }
    ],
    "locationTypes": [
        {
            "documentLocationTypeId": 2572834948,
            "documentId": 529008084,
            "documentAddressId": 1353014511,
            "locationTypeCode": "ShipTo"
        },
        {
            "documentLocationTypeId": 2572834947,
            "documentId": 529008084,
            "documentAddressId": 1353014510,
            "locationTypeCode": "ShipFrom"
        }
    ],
    "summary": [
        {
            "country": "CA",
            "region": "CA",
            "jurisType": "Country",
            "jurisCode": "CA",
            "jurisName": "CANADA",
            "taxAuthorityType": 45,
            "stateAssignedNo": "",
            "taxType": "Output",
            "taxName": "CANADA GST/TPS",
            "rateType": "Standard",
            "taxable": 2985,
            "rate": 0.05,
            "tax": 149.25,
            "taxCalculated": 149.25,
            "nonTaxable": 0,
            "exemption": 0
        },
        {
            "country": "CA",
            "region": "CA",
            "jurisType": "Country",
            "jurisCode": "CA",
            "jurisName": "CANADA",
            "taxAuthorityType": 45,
            "stateAssignedNo": "",
            "taxType": "LandedCost",
            "taxName": "CA LandedCost",
            "rateType": "Standard",
            "taxable": 0,
            "rate": 188.5,
            "tax": 1885,
            "taxCalculated": 1885,
            "nonTaxable": 0,
            "exemption": 0
        },
        {
            "country": "CA",
            "region": "ON",
            "jurisType": "State",
            "jurisCode": "ON",
            "jurisName": "ONTARIO",
            "taxAuthorityType": 45,
            "stateAssignedNo": "",
            "taxType": "Output",
            "taxName": "ONTARIO HST",
            "rateType": "Standard",
            "taxable": 1100,
            "rate": 0.08,
            "tax": 88,
            "taxCalculated": 88,
            "nonTaxable": 0,
            "exemption": 0
        }
    ],
    "parameters": {
        "AvaTax.LandedCost.ShippingMode": "air"
    },
    "messages": [
        {
            "summary": "Shipment meets CA import duty de minimis threshold. Therefore, import duty applies.",
            "details": "",
            "refersTo": "LandedCost",
            "severity": "Success",
            "source": "Avalara.AvaTax.Services.Tax"
        }
    ]
}
```

Another benefit of this refactor is that the native Magento `sales_order_tax` and `sales_order_tax_item` tables will now contain accurate information. Up to this point, the values in this table have not been accurate.

**Comment by LordZardeck** [comment]
We have changed the AvaTax code to pull from the "lines > details" JSON object. To test this, we did the following:

1. Placed an order with the following conditions:
    ![image](https://user-images.githubusercontent.com/1145989/47232268-d7909f00-d394-11e8-8b36-efbbab031ff3.png)
    * Simple product with quantity of 3
    * Configurable product with quantity of 1
    * Discount of 10%
    * Tax in California, which has three jurisdictions
2. Placed the exact same order before and after this code refactor
3. Compared the data in these tables to ensure that the relevant data matched: `sales_order`, `sales_order_item`, `sales_order_tax`, `sales_order_tax_item`. The `sales_order` and `sales_order_item` tables were exactly the same, minus the timestamp/ID changes. The `sales_order_tax` and `sales_order_tax_item` tables had a number of differences, but they were to be expected due to this PR changing how those tables are populated.

The JSON dump of the before/after comparisons are attached to this comment.
[after_code_refactor.json.txt](https://github.com/classyllama/ClassyLlama_AvaTax/files/2496665/after_code_refactor.json.txt)
[before_code_refactor.json.txt](https://github.com/classyllama/ClassyLlama_AvaTax/files/2496666/before_code_refactor.json.txt)

**Comment by erikhansen** [comment]
Closing this issue. This issue was addressed in https://github.com/classyllama/ClassyLlama_AvaTax/pull/197 and will be released with the upcoming 2.0.0 release.
